src/write.c

```diff
@@ -476,15 +480,15 @@ static avifResult avifEncoderAddImageInternal(avifEncoder * encoder,
             }
             item->codec->csOptions = encoder->csOptions;
 
-            if (gridColorID) {
+            if (cellCount > 1) {
                 item->dimgFromID = gridColorID;
-            } else if (!encoder->data->primaryItemID) {
+            } else {
                 encoder->data->primaryItemID = item->id;
             }
```

Comment:
Please see line 460 (line 464 in new code). cellCount > 1 if and only if gridColorID is nonzero.

It is not necessary to check !encoder->data->primaryItemID. If cellCount == 1 (or equivalently gridColorID == 0), then encoder->data->primaryItemID must be 0, and this loop has only one iteration.